packages/langium/src/lsp/completion/completion-provider.ts

```diff
@@ -123,22 +123,43 @@ export class DefaultCompletionProvider implements CompletionProvider {
             }
         };
 
```

Comment:
This method feels a bit complex (or `unübersichtlich`), because of the inserted `if` statements. You could split it into phases, like `collectCompletionContexts`, `completeOnContext` (already there), `mergeCompletionResults`... maybe you find better names.
Code reads better with less branches.

Reply:
I disagree (at least to a certain degree). The nesting is two levels deep at most and most and fairly clear in my opinion. The reason why I'm quite hesitant to split this method is that every part of it requires a lot of offset+ast+document information. We also have a few early exit conditions (such as the one where we perform completion at the start of the document), which become very difficult to handle once we split this into different methods.